File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
from collections import deque

T = int(sys.stdin.readline())

for _ in range(T):
  N, M, K = map(int,sys.stdin.readline().split())
  graph = [[0]*M for _ in range(N)]

  for _ in range(K):
    x,y = map(int,sys.stdin.readline().split())
    graph[x][y] = 1

  dic = [(0,1),(0,-1),(1,0),(-1,0)]

  def bfs(start):
    q = deque([start])
    while(q):
      cx, cy = q.popleft()
      for dx, dy in dic:
        nx = cx+dx
        ny = cy+dy
        if 0<=nx<N and 0<=ny<M and graph[nx][ny] ==1:
          graph[nx][ny] = 2
          q.append([nx,ny])


  answer = 0
  for i in range(N):
    for j in range(M):
      if graph[i][j] == 1:
        graph[i][j] = 2
        bfs([i,j])
        answer +=1

  print(answer)

File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const fs = require('fs')
const input = fs.readFileSync('/dev/stdin','utf8').trim().split('\n')

const T = Number(input.shift().trim())

const dir = [[1,0],[-1,0],[0,1],[0,-1]]

for (let i = 0;i<T;i++){
    let answer = 0
    const [N,M,K] = input.shift().trim().split(' ').map(Number)
    const graph = [...Array(N)].map(()=>Array(M).fill(0))
    for (let j =0;j<K;j++){
        const [x,y] = input.shift().trim().split(' ').map(Number)
        graph[x][y] = 1
    }


    //bfs로 탐색하면서 연결되어있는 1인 곳만 방문하고 방문 처리 2
    function bfs(x,y){
        const q = [[x,y]]
        graph[x][y] = 2
        while(q.length){
            const [cx,cy] = q.shift()
            for (let [dx, dy] of dir){
                let nx = cx+ dx
                let ny = cy+ dy

                if (nx>=0 && nx<N && ny>=0 && ny<M && graph[nx][ny] ===1){
                    graph[nx][ny] = 2
                    q.push([nx,ny])
                }
            }
        }
    }

    function dfs(x,y){
        graph[x][y] = 2
        for (let [dx,dy] of dir){
            let nx = x+dx
            let ny = y+dy
            if (nx>=0 && nx<N && ny>=0 && ny<M && graph[nx][ny] ===1){
                dfs(nx,ny)
            }
        }
    }

    //for문 돌면서 1인 공 방문하고 거기서 퍼지는 거 확인하기 위해 bfs 돌리기
    for (let i=0;i<N;i++){
        for (let j =0;j<M;j++){
            if (graph[i][j] === 1){
                dfs(i,j)
                answer++
            }
        }
    }

    console.log(answer)
}
